### PR TITLE
Issue #6 : Share existing shopping list to a user 

### PR DESCRIPTION
**Pull request by stacy-tech**
## Description

- Implemented a new feature that allows users to invite existing users to share their shopping list using an email ID. 
- Created a form with a labeled input field for entering the user's email ID.
- Structured a `Share List` button that can be activated using both the `mouse` and the `enter` key.
- Added a message to alert the user if the list was successfully shared with the invited user.
- Display an error message if the email doesn't correspond to an existing user in the database. 
- The form correctly identifies and submits valid email addresses.

## Related Issue

 - closes issue #6

## Acceptance Criteria

The ManageList view shows a form that allows the user to enter an email to invite an existing user to a list, in addition to the form that allows them to add items to that list.

- [x] The input that accepts the email has a semantic label element associated with it
- [x] The user can submit this form with both the mouse and the Enter key
- [x]  If the other user exists, the user is alerted that the list was shared
- [x] If the other user does not exist, the user is shown an error message that explains the problem

## Type of Changes

`Enhancement`

### Before

image (1)

### After
image

Screenshot 2024-08-26 at 2 57 58 PM

image

Screenshot 2024-08-26 at 2 59 35 PM

## Testing Steps / QA Criteria

- Sign In using email account.
- Click on the list to be shared. 
- Please note that a user can only share their own list. 
- Navigate to the `Manage List` page.
- Input the recipient's email address and either click on the `Share List` button or press `Enter`.
- If the recipient's email address exists, a success alert will appear.
- An error message is shown if the invited user doesn't exist and if the list has already been shared with the recipient's email address.

### CODE REVIEW

**Review by dterceroparker — APPROVED** [review]
Tests worked for me! Well done! My only question pertains to shareList function. If it is an async function (firebase, line 140), would it make sense to add an await in ManageList? I'm wondering this myself. Great job!

**Review by NickRoccodev11 — APPROVED** [review]

**Comment by NickRoccodev11** [comment]
looks great! I would mention in the testing criteria that a list you share has to be one you created, but all seems to be working well!

**Comment by shuveksha-tuladhar** [comment]
We have updated the testing criteria. Thank you @NickRoccodev11

**Comment by shuveksha-tuladhar** [comment]
> Tests worked for me! Well done! My only question pertains to shareList function. If it is an async function (firebase, line 140), would it make sense to add an await in ManageList? I'm wondering this myself. Great job!

Thank you for the feedback @dterceroparker !  We believe adding await in either Firebase.js file and ManageList.jsx would give the same result. So Stacy and I decided to keep the code as is.

File: src/views/ManageList.jsx [hunk 1]
```diff
@@ -18,6 +21,12 @@ export function ManageList({ listPath }) {
 		setDaysUntilNextPurchase(7);
 	};
 
+	const handleShare = (event) => {
+		event.preventDefault();
+		shareList(listPath, currentUserId, recipientEmail);
```
Comment (jtkabenni) [inline src/views/ManageList.jsx]: shareList is an async function, so I would consider updating the handler function to allow for this, incase you want to add more logic to the function later on that needs to wait on shareList to finish

**Review by jtkabenni — APPROVED** [review]
Looks great! This is not in scope but maybe a feature to consider - what if the list has already been shared with this user? Is there a condition we can check and a different message we would want to show? 

One other thing to consider is how you are handling messaging to your users - you have a couple options here and we already see a few of them in action (e.g native 'required' or email type error messages that are default to the browsers built in functionality). However when I add an invalid email, i get a popup alert. I would think about how to unify messaging across the app so the user experience is similar whenever a user gets a message. 

Like I said, not blockers but just some things to consider!

**Comment by jtkabenni** [comment]
> > Tests worked for me! Well done! My only question pertains to shareList function. If it is an async function (firebase, line 140), would it make sense to add an await in ManageList? I'm wondering this myself. Great job!
> 
> Thank you for the feedback @dterceroparker ! We believe adding await in either Firebase.js file and ManageList.jsx would give the same result. So Stacy and I decided to keep the code as is.

It may seem like the same result but it is slightly different! 👀

File: src/views/ManageList.jsx (hunk 1 above)
Comment (shuveksha-tuladhar) [reply src/views/ManageList.jsx]: @jtkabenni Thank you so much for the suggestions I have refactored the implementation by removing the async and await keywords and instead using the `Resolve` and `Reject` state of Promise when calling the shareList function. This approach allows us to handle the response from the Promise more cleanly and move the alert logic out of the firebase.js file. I believe this makes the code more streamlined and easier to read and maintain. Could you please review these changes when you have a moment? 
Also, informing the rest of the team about this update to ensure everyone is aligned. @stacy-tech @dterceroparker @NickRoccodev11

**Review by jtkabenni — APPROVED** [review]
Looks good!

File: src/views/ManageList.jsx [hunk 2]
```diff
@@ -18,6 +21,18 @@ export function ManageList({ listPath }) {
 		setDaysUntilNextPurchase(7);
 	};
 
+	const handleShare = (event) => {
+		event.preventDefault();
+		shareList(listPath, currentUserId, recipientEmail)
```
Comment (jtkabenni) [inline src/views/ManageList.jsx:26]: this is fine but using async await is also an option here

**Review by jtkabenni — APPROVED** [review]

**Pull request merged**